### DISCUSSION TITLE
What is the result of having multiple rules with the same test in Webpack config?

### DISCUSSION THREAD

**Discussion by SetTrend** [webpack/webpack#15537] [Q&A]
The Webpack documentation on [`module.rules`](https://webpack.js.org/configuration/module/#modulerules) is quite sparse:

> `[Rule]`
>
> An array of Rules which are matched to requests when modules are created. These rules can modify how the module is created. They can apply loaders to the module, or modify the parser.

<br/>

In the [`html-loader` documentation](https://webpack.js.org/loaders/html-loader/), a number of examples exist with two rules testing for the same regular expression, e.g.:

> ```js
> module.exports = {
>   output: {
>     assetModuleFilename: "[name][ext]",
>   },
>   module: {
>     rules: [
>       {
>         test: /\.html$/,
>         type: "asset/resource",
>         generator: {
>           filename: "[name][ext]",
>         },
>       },
>       {
>         test: /\.html$/i,
>         use: ["extract-loader", "html-loader"],
>       },
>     ],
>   },
> };
> ```

<br/>

What is the reason/result/effect/desire of having two rules handling the same file?

Why isn't a single rule being used?

```js
module.exports = {
  output: {
    assetModuleFilename: "[name][ext]",
  },
  module: {
    rules: [{
        test: /\.html$/,
        use: ["extract-loader", "html-loader"],
        type: "asset/resource",
        generator: {
          filename: "[name][ext]",
        },
      }
    ],
  },
};
```

**Comment by Bilge** [comment]
Agree, the documentation does not explain:

- For a given module, can it match multiple rules, or just one?
- If just one, which one matches first, the first declared or the last declared?
- If multiple, in which order are they applied?

Note, this is NOT the same question as loader order, which is well defined; this is a question of _rule order_, which is left to the imagination.

**Reply by Bilge** [reply]
FWIW I asked Copilot, and it says multiple rules can match (unless you use `oneOf`), and they are applied from top to bottom. But it's AI so YMMV lol.

>When multiple rules match a module in webpack, all matching rules are applied in the order they are defined in the configuration. This means the first matching rule is applied first, followed by the second, and so on.

**Reply by Bilge** [reply]
Empirically I have confirmed Copilot was actually correct.
